src/systems/systems/src/air_conditioning/acs_controller.rs

```diff
@@ -0,0 +1,1808 @@
+            zone_selected_temperature: ThermodynamicTemperature::new::<degree_celsius>(24.),
+            zone_measured_temperature: ThermodynamicTemperature::new::<degree_celsius>(24.),
+            pid_controller,
+        }
+    }
+
+    fn update(
+        &mut self,
+        context: &UpdateContext,
+        acs_overhead: &AirConditioningSystemOverhead,
+        pack_flow: &impl PackFlow,
+        pressurization: &impl CabinAltitude,
+    ) {
+        if acs_overhead
+            .selected_cabin_temperatures()
+            .contains_key(&self.zone_id)
+        {
+            self.zone_selected_temperature =
+                acs_overhead.selected_cabin_temperatures()[&self.zone_id];
+        }
+        self.duct_demand_temperature =
+            if pack_flow.pack_flow() < MassRate::new::<kilogram_per_second>(0.01) {
+                // When there's no pack flow, duct temperature is mostly determined by cabin recirculated air and ambient temperature
+                ThermodynamicTemperature::new::<degree_celsius>(
+                    0.8 * self.zone_measured_temperature.get::<degree_celsius>()
+                        + 0.2 * context.ambient_temperature().get::<degree_celsius>(),
+                )
+            } else {
+                self.calculate_duct_temp_demand(context, pressurization)
+            };
+    }
+
+    fn calculate_duct_temp_demand(
+        &mut self,
+        context: &UpdateContext,
+        pressurization: &impl CabinAltitude,
+    ) -> ThermodynamicTemperature {
+        let altitude_correction: f64 =
+            pressurization.cabin_altitude().get::<foot>() * Self::K_ALTITUDE_CORRECTION;
+        let corrected_selected_temp: f64 =
+            self.zone_selected_temperature.get::<kelvin>() + altitude_correction;
+
+        self.pid_controller
+            .change_max_output(self.calculate_duct_temp_upper_limit().get::<kelvin>());
+        self.pid_controller
+            .change_min_output(self.calculate_duct_temp_lower_limit().get::<kelvin>());
+        self.pid_controller.change_setpoint(corrected_selected_temp);
+
+        let duct_demand_limited: f64 = self.pid_controller.next_control_output(
+            self.zone_measured_temperature.get::<kelvin>(),
+            Some(context.delta()),
+        );
+        ThermodynamicTemperature::new::<kelvin>(duct_demand_limited)
+    }
+
+    fn calculate_duct_temp_upper_limit(&self) -> ThermodynamicTemperature {
+        if self.zone_measured_temperature > ThermodynamicTemperature::new::<degree_celsius>(19.) {
+            ThermodynamicTemperature::new::<kelvin>(Self::UPPER_DUCT_TEMP_LIMIT_LOW)
+        } else if self.zone_measured_temperature
+            < ThermodynamicTemperature::new::<degree_celsius>(17.)
+        {
+            ThermodynamicTemperature::new::<kelvin>(Self::UPPER_DUCT_TEMP_LIMIT_HIGH)
+        } else {
+            let interpolation =
+                (Self::UPPER_DUCT_TEMP_LIMIT_LOW - Self::UPPER_DUCT_TEMP_LIMIT_HIGH) / (19. - 17.)
```

Comment:
**thought**: It seems 17 and 19 have some significance in this code. Maybe a local const would be useful?

As a general remark with magic numbers used in this code (and the functions around it): try to give them a name where it isn't immediately obvious what the number represents.

Reply:
Eliminated magic numbers.

Reply:
Eliminated magic numbers.

Reply:
Twice even! 😀